File: rd-net/RdFramework/Impl/RdEntitiesRegistrar.cs
using System.Collections.Generic;
using JetBrains.Diagnostics;
using JetBrains.Lifetimes;
using JetBrains.Rd.Base;
using JetBrains.Rd.Util;

namespace JetBrains.Rd.Impl;

public class RdEntitiesRegistrar
{
  private readonly Dictionary<RdId, IRdDynamic> myMap = new();

  internal void Register(Lifetime lifetime, RdId rdId, IRdDynamic dynamic)
  {
    Assertion.Assert(!rdId.IsNil);

    myMap.BlockingAddUnique(lifetime, myMap, rdId, dynamic);
  }

  public bool TryGetEntity(RdId rdId, out IRdDynamic entity)
  {
    lock (myMap)
    {
      return myMap.TryGetValue(rdId, out entity);
    }
  }
}

File: rd-kt/rd-framework/src/main/kotlin/com/jetbrains/rd/framework/RdEntitiesRegistrar.kt
package com.jetbrains.rd.framework

import com.jetbrains.rd.util.addUnique
import com.jetbrains.rd.util.lifetime.Lifetime
import java.util.concurrent.ConcurrentHashMap

class RdEntitiesRegistrar {
    private val map = ConcurrentHashMap<RdId, IRdDynamic>()

    internal fun register(lifetime: Lifetime, rdId: RdId, dynamic: IRdDynamic) {
        require(!rdId.isNull)
        map.addUnique(lifetime, rdId, dynamic)
    }

    fun tryGetDynamic(rdId: RdId): IRdDynamic? = map[rdId]
}